test-system/src/lib.rs

```diff
@@ -135,8 +138,55 @@ macro_rules! test_runtest {
         test_declare!($n);
         let res: TestResult = $b;
         match res {
-            TestResult::Pass => test_pass!($n),
-            TestResult::Fail => test_fail!($n),
+            Ok(()) => test_pass!($n),
+            Err(TestFailure::Fail) => test_fail!($n),
+            Err(TestFailure::FailedAt(s)) => test_fail!($n, s),
+        }
+    };
+}
+
+/// Test if an expression is true, with a name for the test
+/// and a boolean to carry results from multiple test
+/// ```
+/// use test_system::*;
+/// let mut results = true;
+/// test_result_true(1 == 1, "first", results);
+/// test_result_true(2 == 2, "second", results);
+///
+/// ```
+#[macro_export]
+macro_rules! test_result_true {
+    ($e:expr, $n: expr, $f:expr) => {
+        test_declare!($n);
+        let res: bool = $e;
+        $f = $f && res;
+        if res {
+            test_pass!($n);
+        } else {
+            test_fail!($n)
+        }
+    };
+}
```

Comment:
instead of accumulating test results in a `bool`, can this instead evaluate to a `TestResult` that we can use the `?` operator on?

e.g.

```
fn test_fn() -> TestResult {
    test_result_true!(1 == 1, "foo")?;
    Ok(())
}
```

Reply:
fixed. (btw I don't think you can use the ? operator on a macro)

Reply:
There shouldn't be inherent limitation on the use of the ? operator inside macros, and long as the encompassing scope can handle it in conjunction with ?. 

For example, the macro itself can return the `Result<>` as an expression, meaning that we can do something like:

```
#[macro_export]
macro_rules! test_result_true {
    ($e:expr, $n: expr) => {
       {
        test_declare!($n);
        let res: bool = $e;
        if res {
            test_pass!($n);
            Ok(())
        } else {
            test_fail!($n);
           Err(())
        }
    }
    }
}

test_result_true!(1 == 1, "One equals")?;
test_result_true(2 == 2, "Two equals")?
```

Anyway, I am not sure whether it's worth modifying what you have already implemented since it seems to have a similar effect.